In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# BERT

Here is a page listing all available BERT models on TensorFlow Hub that you can download and use:

- [BERT Models on TensorFlow Hub](https://tfhub.dev/google/collections/bert/1)

For information on the basic uncased BERT model:

- [BERT Model Details](https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4)

This model uses:
- L=12 hidden layers (i.e., Transformer blocks)
- A hidden size of H=768
- A=12 attention heads

It has been pre-trained for English on the Wikipedia and BooksCorpus.


In [2]:
import tensorflow_hub as hub
import tensorflow_text as text

import warnings
warnings.filterwarnings('ignore')

2024-07-25 02:45:25.214747: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-07-25 02:45:25.214939: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-07-25 02:45:25.394961: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [3]:
preprocess_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3"
encoder_url = "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4"

In [4]:
bert_preprocess_model = hub.KerasLayer(preprocess_url)

In [5]:
text_test = ['nice movie indeed','I love python programming']
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_mask', 'input_type_ids', 'input_word_ids'])

In [6]:
text_preprocessed['input_mask']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [7]:
text_preprocessed['input_type_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [8]:
text_preprocessed['input_word_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[  101,  3835,  3185,  5262,   102,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0, 

**101 --> CLS Token**

**102 --> SEP Token**

BERT uses the CLS token as a special token at the beginning of each sentence, while the SEP token is used to separate two sentences or to denote the end of a single sentence.


In [9]:
bert_model = hub.KerasLayer(encoder_url)

In [10]:
bert_results = bert_model(text_preprocessed)

In [11]:
bert_results.keys()

dict_keys(['sequence_output', 'default', 'pooled_output', 'encoder_outputs'])

In [12]:
bert_results['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[ 0.07292047,  0.08567817,  0.14476831, ..., -0.09677088,
          0.08722127,  0.07711085],
        [ 0.1783942 , -0.19006078,  0.5034952 , ..., -0.05869824,
          0.32717073, -0.15578493],
        [ 0.18701449, -0.43388748, -0.4887515 , ..., -0.15502778,
          0.00145182, -0.24471016],
        ...,
        [ 0.12083091,  0.12884252,  0.46453542, ...,  0.0737552 ,
          0.17441946,  0.16522075],
        [ 0.07967882, -0.01190652,  0.502254  , ...,  0.13777725,
          0.21002197,  0.00624588],
        [-0.07212701, -0.28303462,  0.59033376, ...,  0.475519  ,
          0.16668482, -0.08920349]],

       [[-0.0790059 ,  0.36335096, -0.21101595, ..., -0.17183761,
          0.16299714,  0.6724268 ],
        [ 0.27883548,  0.43716297, -0.35764822, ..., -0.04463654,
          0.38315183,  0.5887981 ],
        [ 1.2037671 ,  1.0727016 ,  0.48408663, ...,  0.24921048,
          0.407309  ,  0.40481755],
        ...,

In [13]:
bert_results['pooled_output']

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.79177403, -0.21411912,  0.49769545, ...,  0.24465235,
        -0.47334474,  0.81758696],
       [-0.91712314, -0.47935176, -0.7865697 , ..., -0.6175175 ,
        -0.71026856,  0.921843  ]], dtype=float32)>

In [14]:
len(bert_results['encoder_outputs'])

12

Since we are using the BERT base model, which has 12 encoder layers, the length of the `encoder_output` is 12. 

Read more about the purpose of each element in the model [here](https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4).


We can see below that the last element of `encoder_outputs` is essentially the `sequence_output`.


In [15]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]]])>

In [16]:
bert_results['encoder_outputs']

[<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
 array([[[ 0.1290144 ,  0.00644745, -0.03614967, ...,  0.04999624,
           0.06149195, -0.02657542],
         [ 1.1753386 ,  1.2140785 ,  1.1569982 , ...,  0.11634357,
          -0.3585537 , -0.40490186],
         [ 0.03859041,  0.5386997 , -0.2108978 , ...,  0.21858197,
           0.7260168 , -1.1158606 ],
         ...,
         [-0.07587015, -0.25421894,  0.7075512 , ...,  0.50541997,
          -0.1887868 ,  0.15028352],
         [-0.16066599, -0.28089687,  0.5759707 , ...,  0.5275855 ,
          -0.11141382,  0.02887543],
         [-0.04428157, -0.20279568,  0.59093547, ...,  0.8133835 ,
          -0.39075816, -0.02601742]],
 
        [[ 0.18903585,  0.02752551, -0.06513745, ..., -0.00620203,
           0.15053892,  0.0316545 ],
         [ 0.5916147 ,  0.7589138 , -0.07240658, ...,  0.61903954,
           0.8292891 ,  0.16161972],
         [ 1.4460826 ,  0.44602674,  0.40990272, ...,  0.48255897,
           0.6269117 ,  0.13